### PR TITLE
[python] reset storages in early stopping callback after finishing training

### PR DESCRIPTION
**Pull request by StrikerRUS**
Right now `early_stopping()` callback cannot be used with scikit-learn's `GridSearchCV` tool due to that storages for metric results are not cleaned after early stopping occurs. This results in **comparing current results with the best iteration for the first grid iteration for the first fold**.

This was spotted by our `test_grid_search` test
https://github.com/microsoft/LightGBM/blob/fe535a0e9dfb251fcc85dd33fb0bf193d022fe90/tests/python_package_test/test_sklearn.py#L276
which fails with early stopping setup passed via callback with the following error
```
>       assert grid.best_estimator_.best_score_['valid_0']['multi_logloss'] < 0.25
E       assert 0.5915701709051111 < 0.25

../tests/python_package_test/test_sklearn.py:318: AssertionError
```

This PR proposes resetting all global variables in the `_init()` helper function and determining whether initialization is required by the special variable `inited` instead of checking `cmp_op` global variable for emptiness, which results right now in that `_init()` is called only once during the whole grid search routine.

Extracted from #4846.

### CODE REVIEW

File: python-package/lightgbm/callback.py [hunk 1]
```diff
@@ -283,13 +293,15 @@ def _init(env: CallbackEnv) -> None:
...
         nonlocal best_score_list
+        nonlocal inited
         if env.iteration == env.end_iteration - 1:
             if verbose:
                 best_score_str = '\t'.join([_format_eval_result(x) for x in best_score_list[i]])
                 _log_info('Did not meet early stopping. '
                           f'Best iteration is:\n[{best_iter[i] + 1}]\t{best_score_str}')
                 if first_metric_only:
                     _log_info(f"Evaluated only: {eval_name_splitted[-1]}")
+            inited = False
```
Comment (shiyu1994) [inline python-package/lightgbm/callback.py:304]: It seems that `inited` is reset to `False` only when early stopping is not triggered? Because
1. `inited` is reset to `False` only when `env.iteration == env.end_iteration - 1`
2. `_final_iteration_check` is not called when early stopping is triggered according to
https://github.com/microsoft/LightGBM/blob/00f87c52ec24b1f8358c230646685459e31b60b8/python-package/lightgbm/callback.py#L306-L327

**Review by shiyu1994 — COMMENTED** [review]
Thank you for working on this. Just leave a comment for now since I'm not sure the `inited` is reset correctly.

File: python-package/lightgbm/callback.py (hunk 1 above)
Comment (StrikerRUS) [reply python-package/lightgbm/callback.py:304]: @shiyu1994 Thanks for your review!

> It seems that `inited` is reset to `False` only when early stopping is not triggered?

I believe `inited` is reset to `False` when early stopping is triggered here
https://github.com/microsoft/LightGBM/blob/dff622c26aceb5bc602a6279d4a45d7e99d389e5/python-package/lightgbm/callback.py#L339-L340
Triggering early stopping means to raise `EarlyStopException` and `inited` is reset to `False` right before that line.

Also, if it is not reset, the test `test_grid_search()` will fail because we use constant custom evaluation metric there which force early stopping to happen at `stopping_rounds` iteration because there is no improvement after the first iteration.

**Comment by shiyu1994** [comment]
@StrikerRUS Thanks for the explanation. Sorry that I did not notice line 339.

**Review by shiyu1994 — APPROVED** [review]
The changes LGTM. Thank you.

**Review by jameslamb — APPROVED** [review]
thanks very much for the explanation! I think this fix makes sense.

**Comment by PhillipMaire** [comment]
is this going to be merged to master? having a working grid search and other related packaged like OPTUNA work with light GBM is a must for many people and without this update they still don't work properly with light GBM!! from [OPTUNA #3145](https://github.com/optuna/optuna/issues/3145#issuecomment-1118012232) "LightGBM 3.3.2, which has been released yesterday, does not include the change. We need to wait for the fix..." and recently [OPTUNA #3625](https://github.com/optuna/optuna/issues/3625)

**Comment by jameslamb** [comment]
> is this going to be merged to `master`

This PR was merged on December 9, 2021. I think you mean "when will this be released to package managed like PyPI".

Due to a lack of maintainer activity, it will probably still be several months until the next release of LightGBM. You can subscribe to #5153 for updates on the next release, and even comment on the linked issues if there are any you'd like to contribute, to move the project closer to that release.

**Comment by PhillipMaire** [comment]
ahh yes thank you for clarifying this! I installed it on my [test notebook](https://colab.research.google.com/drive/1LigmaNSoK_FFi1BYTwhaDHZEBiLZskz0#scrollTo=wZQXbecNuArq) and it seems to run fine but I get an error with OPTUNA which I mentioned to them. I installed using the following in case anyone else needs this 
```
!git clone --recursive https://github.com/microsoft/LightGBM.git
!cd LightGBM/python-package && python setup.py install
```
